From 4e3c30905a877f27c896c06cb260bf3aaf3394f3 Mon Sep 17 00:00:00 2001
From: hienhuynh <hien.huynh.px@renesas.com>
Date: Thu, 23 Jun 2022 11:06:12 +0700
Subject: [PATCH 04/27] pinctrl: pinctrl-rzg2l: add support Slew Rate pin
 configuration setting

This commit is created to support Slew Rate pin configuration setting
for RZ/G2L Series.
To support this feature, please add "slew-rate" property in pinctrl
node for each module:
	slew-rate = <0>
or
	slew-rate = <1>
where	0: Slow
	1: Fast (Default value)

Signed-off-by: hienhuynh <hien.huynh.px@renesas.com>
---
 drivers/pinctrl/renesas/pinctrl-rzg2l.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/pinctrl/renesas/pinctrl-rzg2l.c b/drivers/pinctrl/renesas/pinctrl-rzg2l.c
index bfb5ca699725..ba9b4d664be2 100644
--- a/drivers/pinctrl/renesas/pinctrl-rzg2l.c
+++ b/drivers/pinctrl/renesas/pinctrl-rzg2l.c
@@ -92,6 +92,7 @@
 #define PFC(n)			(0x0400 + 0x40 + (n) * 4)
 #define PIN(n)			(0x0800 + 0x10 + (n))
 #define IOLH(n)			(0x1000 + (n) * 8)
+#define SR(n)			(0x1400 + (n) * 8)
 #define IEN(n)			(0x1800 + (n) * 8)
 #define ISEL(n)			(0x2C00 + 0x80 + (n) * 8)
 #define PWPR			(0x3014)
@@ -113,6 +114,7 @@
 #define ETH_PVDD_MASK		0x03
 #define PFC_MASK		0x07
 #define IEN_MASK		0x01
+#define SR_MASK			0x01
 #define IOLH_MASK		0x03
 
 #define PM_INPUT		0x1
@@ -722,6 +724,14 @@ static int rzg2l_pinctrl_pinconf_get(struct pinctrl_dev *pctldev,
 		break;
 	}
 
+	case PIN_CONFIG_SLEW_RATE: {
+		if (!(cfg & PIN_CFG_SR))
+			return -EINVAL;
+
+		arg = rzg2l_read_pin_config(pctrl, SR(port_offset), bit, SR_MASK);
+		break;
+	}
+
 	default:
 		return -ENOTSUPP;
 	}
@@ -850,6 +860,18 @@ static int rzg2l_pinctrl_pinconf_set(struct pinctrl_dev *pctldev,
 			break;
 		}
 
+		case PIN_CONFIG_SLEW_RATE: {
+			unsigned int arg =
+					pinconf_to_config_argument(_configs[i]);
+
+			if (!(cfg & PIN_CFG_SR))
+				return -EINVAL;
+
+			rzg2l_rmw_pin_config(pctrl, SR(port_offset),
+					     bit, SR_MASK, !!arg);
+			break;
+		}
+
 		default:
 			return -EOPNOTSUPP;
 		}
-- 
2.17.1

